src/ape/api/projects.py

```diff
@@ -0,0 +1,229 @@
+
+    @property
+    @abstractmethod
+    def version(self) -> str:
+        """
+        The version of the project.
+        """
+
+    @property
+    def manifest_cachefile(self) -> Path:
+        """
+        The path to the project's cached manifest. The manifest
+        is a cached file representing the project and is useful
+        for sharing, such as uploading to IPFS.
+
+        Returns:
+            pathlib.Path
+        """
+
+        file_name = self.name or "__local__"
+        return self.cache_folder / (file_name + ".json")
+
+    @property
+    def cached_manifest(self) -> Optional[PackageManifest]:
+        """
+        The cached ``PackageManifest``.
+        If nothing has been compiled, then no manifest will exist, and
+        this will return ``None``.
+        """
+
+        if self.cache_folder.exists() and not self.manifest_cachefile.exists():
+            shutil.rmtree(self.cache_folder)
+
+        self.cache_folder.mkdir(exist_ok=True, parents=True)
+
+        if not self.manifest_cachefile.exists():
+            return None
+
+        def handle_corrupt_manifest():
+            logger.warning(f"Existing manifest file for '{self.name}' corrupted.")
+            self.manifest_cachefile.unlink()
+
+        try:
+            manifest_dict = json.loads(self.manifest_cachefile.read_text())
+            if not isinstance(manifest_dict, dict) or "manifest" not in manifest_dict:
+                handle_corrupt_manifest()
+                return None
+
+        except json.JSONDecodeError:
+            handle_corrupt_manifest()
+            return None
+
+        return PackageManifest(**manifest_dict)
+
+    @classmethod
+    def _create_manifest(
+        cls,
+        sources: Collection[Path],
+        contracts_path: Path,
+        contract_types: Dict[str, ContractType],
+        name: Optional[str] = None,
+        version: Optional[str] = None,
+        initial_manifest: Optional[PackageManifest] = None,
+    ) -> PackageManifest:
+        manifest = initial_manifest or PackageManifest()
+
+        if name:
+            manifest.name = PackageName(name.lower().replace("_", "-"))
+
+        if version:
+            manifest.version = version
+
+        manifest.sources = cls._create_source_dict(sources, contracts_path)
+        manifest.contract_types = contract_types
+        return PackageManifest(**manifest.dict())  # Re-validate
+
+    @classmethod
+    def _create_source_dict(
+        cls, contracts_paths: Collection[Path], base_path: Path
+    ) -> Dict[str, Source]:
+        return {
+            str(get_relative_path(source, base_path)): Source(  # type: ignore
+                checksum=Checksum(  # type: ignore
+                    algorithm="md5",
+                    hash=compute_checksum(source.read_bytes()),
+                ),
+                urls=[],
+                content=source.read_text(),
+            )
+            for source in contracts_paths
+        }
+
+
+class DependencyAPI(ProjectAPI):
```

Comment:
Was hoping this could be separate from `ProjectAPI`

Reply:
If we make them separate, we have to be able to re-create new compilers, config object, etc. Like a whole new ape instance is needed or else nothing works.

Reply:
this is why I had to make the `compile_external_project` method. Because otherwise, we need to re-initialize a new `CompilerManager` object

Reply:
Will get this to work the way you want.